solr/core/src/java/org/apache/solr/update/TransactionLog.java

```diff
@@ -527,18 +562,18 @@ public Object lookup(long pos) {
       // make sure any unflushed buffer has been flushed
       synchronized (this) {
         // TODO: optimize this by keeping track of what we have flushed up to
-        fos.flushBuffer();
+        fos.flush();
```

Comment:
there is a difference between `FastOutputStream` `flush` and `flushBuffer` methods. have you verified this does not introduce some performance regression due to calling the flush method?

Reply:
Right, the methods are different. But the choice here of using flushBuffer is questionable.
(in addition, perf benchmark confirms no difference)

Line 565, fos.flushBuffer() followed by opening an input stream on the channel to read it. It works fine because fos delegates to a Channel output stream with no buffering. When fos.flushBuffer() writes to its delegate out, it writes directly to the channel. So calling fos.flush() which calls the delegate out.flush() is a no op. If now we want to support any delegate output stream with OutputStreamOpener.open(), then we need to call fos.flush().

Line 615, fos.flushBuffer() is called in the finish method to finish the transaction log. We should call flush() to ensure all intermediate output buffers are flushed.

Line 735, fos.flushBuffer() is called just before reading the corresponding input stream fis. Again, this works because the current delegate output stream has no buffering.

Line 876, fos.flushBuffer() is called just before asserting fos.size == channel.size(). Again, this works because the current delegate output stream has no buffering.

Reply:
this makes sense to me. thank you for all the details.